compiler/rustc_smir/src/rustc_smir/mod.rs

```diff
@@ -1205,3 +1213,66 @@ impl<'tcx> Stable<'tcx> for ty::TraitRef<'tcx> {
         TraitRef { def_id: rustc_internal::trait_def(self.def_id), args: self.args.stable(tables) }
     }
 }
+
+impl<'tcx> Stable<'tcx> for ty::Generics {
+    type T = stable_mir::ty::Generics;
+
+    fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
+        use stable_mir::ty::Generics;
+        let def_id = self.parent.map(|did| tables.generic_def(did)).or(None);
+
+        let mut param_def_id_to_index: FxHashMap<GenericDef, u32> = Default::default();
+
+        #[allow(rustc::potential_query_instability)]
+        for (def_id, index) in self.param_def_id_to_index.iter() {
+            param_def_id_to_index.insert(tables.generic_def(*def_id), *index);
+        }
+
+        let mut params: Vec<GenericParamDef> = Vec::new();
+
+        for param in self.params.iter() {
+            params.push(param.stable(tables));
+        }
+
+        Generics {
+            parent: def_id,
+            parent_count: self.parent_count,
+            params,
+            param_def_id_to_index: param_def_id_to_index,
+            has_self: self.has_self,
+            has_late_bound_regions: Some(opaque(&self.has_late_bound_regions)),
+            host_effect_index: self.host_effect_index,
+        }
+    }
+}
```

Comment:
I'd implement this more like:

```rust
    fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
        use stable_mir::ty::Generics;

        let params: Vec<_> = self.params.iter().map(|param| {
            param.stable(tables)
        }).collect();
        let param_def_id_to_index = params.iter().map(|param| (param.def_id, param.index)).collect();

        Generics {
            parent: self.parent.map(|did| tables.generic_def(did)),
            parent_count: self.parent_count,
            params,
            param_def_id_to_index,
            has_self: self.has_self,
            has_late_bound_regions: self.has_late_bound_regions.as_ref().map(|late_bound_regions| late_bound_regions.stable(tables)),
            host_effect_index: self.host_effect_index,
        }
    }
```

Then you can remove `FxHashMap` requirement and the dependency of `rustc_data_structures`. Would need to implement stable for Span and would need `GenericDef` to be `Copy`, just make all these Defs copy. Otherwise just call `opaque` directly for now.